src/server/pfs/server/driver.go

```diff
@@ -793,34 +797,40 @@ func (d *driver) getCompactedDiffFileSet(ctx context.Context, commit *pfs.Commit
 // The ProjectInfo provided to the closure is repurposed on each invocation, so it's the client's responsibility to clone the ProjectInfo if desired
 func (d *driver) listProject(ctx context.Context, cb func(*pfs.ProjectInfo) error) error {
 	authIsActive := true
-	projectInfo := &pfs.ProjectInfo{}
-	return errors.Wrap(d.projects.ReadOnly(ctx).List(projectInfo, col.DefaultOptions(), func(string) error {
-		if authIsActive {
-			resp, err := d.env.AuthServer.GetPermissions(ctx, &auth.GetPermissionsRequest{Resource: projectInfo.GetProject().AuthResource()})
-			if err != nil {
-				if errors.Is(err, auth.ErrNotActivated) {
-					// Avoid unnecessary subsequent Auth API calls.
-					authIsActive = false
-					return cb(projectInfo)
+	return errors.Wrap(dbutil.WithTx(ctx, d.env.DB, func(ctx context.Context, tx *pachsql.Tx) error {
+		projIter, err := coredb.ListProject(ctx, tx)
+		if err != nil {
+			return err
+		}
+		return stream.ForEach[*pfs.ProjectInfo](ctx, projIter, func(proj *pfs.ProjectInfo) error {
+			if authIsActive {
+				resp, err := d.env.AuthServer.GetPermissions(ctx, &auth.GetPermissionsRequest{Resource: proj.GetProject().AuthResource()})
+				if err != nil {
+					if errors.Is(err, auth.ErrNotActivated) {
+						// Avoid unnecessary subsequent Auth API calls.
+						authIsActive = false
```

Comment:
Something that's weird to me is that Pachyderm transactions have some internal state that knows if auth is activated or not, it seems weird that we're skipping that here.  But I guess the old code did it.

Reply:
So there's a couple of different ways to determine whether auth is active. The current "best practice" is just to assume auth is active, make any auth call and see if you get back a `ErrNotActivated`. Alternatively, we can just look at the db and see if a role binding exists for the cluster resource. I think the prior is better.

However, as you can see here, this is an example of doing IO within a SQL transaction, which is the constraint we relaxed as part of our design discussion around transactions.

Final point is that this code is a bit scary, since `GetPermissions` is not transactional. We are basically hoping that permissions don't change that much during the life time of this transactional list projects. I think for now, this is a safe assumption.